### PR TITLE
added arbitrary item sorting

### PR DESCRIPTION
**Pull request by K-91**

### CODE REVIEW

**Comment by K-91** [comment]
I have read the CLA Document and I hereby sign the CLA

File: ModComponents/ItemSorting.cs [hunk 1]
```diff
@@ -44,24 +43,26 @@ public static List<EquipmentIndex> sortItems(List<EquipmentIndex> equipmentList,
         }
 
         private delegate IOrderedEnumerable<ItemIndex> DirectSorter(IOrderedEnumerable<ItemIndex> order);
-        private delegate bool ItemFilter(ItemIndex item);
-        private delegate T ItemSorter<T>(IOrderedEnumerable<ItemIndex> order, Inventory inventory, ItemIndex item);
+        private delegate bool ItemFilter(ItemIndex item, ItemIndex idx = ItemIndex.None);
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:46]: Why did you add a second ItemIndex variable?

File: ModComponents/ItemSorting.cs [hunk 2]
```diff
@@ -193,10 +196,44 @@ public static List<ItemIndex> sortItems(List<ItemIndex> itemList, Inventory inve
             bool filtering = false;
             bool tierSelect = false;
             bool tierReversed = false;
+            bool nameParsing = false;
+            bool nameReverse = false;
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:200]: probably make tierReversed universal instead of making one for each? unless it's somehow not possible but I figure it should only be used by one at a time.

File: ModComponents/ItemSorting.cs [hunk 3]
```diff
@@ -193,10 +196,44 @@ public static List<ItemIndex> sortItems(List<ItemIndex> itemList, Inventory inve
...
+            StringBuilder parsedName = new StringBuilder();
             ItemFilter nextFilter = null;
             IOrderedEnumerable<ItemIndex> finalOrder = itemList.OrderBy(a => 1);
             foreach (char c in sortOrder.ToCharArray())
             {
+                if (c == '<' || c == '>')  // Start Named Item Parsing
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:206]: this should probably be part of the switch case below

File: ModComponents/ItemSorting.cs [hunk 4]
```diff
@@ -193,10 +196,44 @@ public static List<ItemIndex> sortItems(List<ItemIndex> itemList, Inventory inve
...
+                    if (c == '(')
+                    {
+                        continue;
+                    }
+                    if (c == ')')
+                    {
+
+                        ItemIndex indexed = ItemCatalog.FindItemIndex(parsedName.ToString());
+                        if (indexed != ItemIndex.None)
+                        {
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:223]: This should probably be creating a ItemSorter method here that uses the ItemIndex, not add an extra idx value to the SortStep type.

**Review by xoxfaby — CHANGES_REQUESTED** [review]
don't change the whole SortStep struct, just create a method there

**Comment by K-91** [comment]
I've made the requested changes

File: ModComponents/ItemSorting.cs [hunk 5]
```diff
@@ -254,9 +283,15 @@ public static List<ItemIndex> sortItems(List<ItemIndex> itemList, Inventory inve
 
                 switch (c)
                 {
+                    case '<':
+                    case '>':
+                        nameParsing = true;
+                        tierReversed = (c == '<');
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:288]: maybe rename tierReversed to just sortReversed or reversed or something since it's used by multiple parsers now

File: ModComponents/ItemSorting.cs [hunk 6]
```diff
@@ -47,21 +46,22 @@ public static List<EquipmentIndex> sortItems(List<EquipmentIndex> equipmentList,
...
 
-        private static ItemFilter scrapFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Scrap);
-        private static ItemFilter damageFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Damage);
-        private static ItemFilter healingFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Healing);
-        private static ItemFilter utilityFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Utility);
-        private static ItemFilter onKillEffectFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.OnKillEffect);
-        private static ItemFilter equipmentRelatedFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.EquipmentRelated);
-        private static ItemFilter sprintRelatedFilter = item => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.SprintRelated);
-        private static Dictionary<Char, ItemFilter> tierFilters = new Dictionary<char, ItemFilter>()
+
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs]: y

File: ModComponents/ItemSorting.cs [hunk 7]
```diff
@@ -47,21 +46,22 @@ public static List<EquipmentIndex> sortItems(List<EquipmentIndex> equipmentList,
...
+        private static ItemFilter scrapFilter = (item) => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Scrap);
+        private static ItemFilter damageFilter = (item) => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Damage);
+        private static ItemFilter healingFilter = (item) => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Healing);
+        private static ItemFilter utilityFilter = (item) => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.Utility);
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:52]: u

File: ModComponents/ItemSorting.cs [hunk 8]
```diff
@@ -47,21 +46,22 @@ public static List<EquipmentIndex> sortItems(List<EquipmentIndex> equipmentList,
...
+        private static ItemFilter sprintRelatedFilter = (item) => ItemCatalog.GetItemDef(item).ContainsTag(ItemTag.SprintRelated);
+        private static Dictionary<char, ItemFilter> tierFilters = new Dictionary<char, ItemFilter>()
         {
-            { '1', item => ItemCatalog.GetItemDef(item).tier == ItemTier.Tier1 || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidTier1 },
-            { '2', item => ItemCatalog.GetItemDef(item).tier == ItemTier.Tier2 || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidTier2 },
-            { '3', item => ItemCatalog.GetItemDef(item).tier == ItemTier.Tier3 || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidTier3 },
-            { 'L', item => ItemCatalog.GetItemDef(item).tier == ItemTier.Lunar },
-            { 'B', item => ItemCatalog.GetItemDef(item).tier == ItemTier.Boss || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidBoss },
-            { 'N', item => ItemCatalog.GetItemDef(item).tier == ItemTier.NoTier},
+            { '1', (item) => ItemCatalog.GetItemDef(item).tier == ItemTier.Tier1 || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidTier1 },
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:58]: do

File: ModComponents/ItemSorting.cs [hunk 9]
```diff
@@ -47,21 +46,22 @@ public static List<EquipmentIndex> sortItems(List<EquipmentIndex> equipmentList,
...
+            { '2', (item) => ItemCatalog.GetItemDef(item).tier == ItemTier.Tier2 || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidTier2 },
+            { '3', (item) => ItemCatalog.GetItemDef(item).tier == ItemTier.Tier3 || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidTier3 },
+            { 'L', (item) => ItemCatalog.GetItemDef(item).tier == ItemTier.Lunar },
+            { 'B', (item) => ItemCatalog.GetItemDef(item).tier == ItemTier.Boss || ItemCatalog.GetItemDef(item).tier == ItemTier.VoidBoss },
+            { 'N', (item) => ItemCatalog.GetItemDef(item).tier == ItemTier.NoTier},
```
Comment (xoxfaby) [inline ModComponents/ItemSorting.cs:63]: dis

**Review by xoxfaby — CHANGES_REQUESTED** [review]
just a few more touch ups <3
